crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+        }
+        todo!()
+    }
+
+    /// read out the current publish state
+    pub fn analysis(&self) -> &Value {
+        self.analysis.deref()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PublishError {
+    #[error("Failed to load metadata for the package '{0}' in '{1}': {2}")]
+    DrvMetadata(AttrPath, PublishRef, NixCommandLineRunJsonError),
+
+    #[error("Failed to load metadata for flake '{0}': {1}")]
+    FlakeMetadata(PublishRef, NixCommandLineRunJsonError),
+}
+
+type PublishResult<T> = Result<T, PublishError>;
+
+/// Publishable FlakeRefs
+///
+/// `publish` modifies branches of the source repository.
+/// Thus we can only publish to repositories in (remote*) git repositories.
+/// This enum represents the subset of flakerefs we can use,
+/// so we can avoid parsing and converting flakerefs within publish.
+/// [GitRef<protocol::File>] should in most cases be resolved to a remote type.
+#[derive(PartialEq, Eq, Clone, Debug, Display)]
+pub enum PublishRef {
+    Ssh(GitRef<protocol::SSH>),
+    Https(GitRef<protocol::HTTPS>),
+    // File(GitRef<protocol::File>),
+}
+
+impl PublishRef {
+    /// extract an url for cloning with git
+    fn clone_url(&self) -> String {
+        match self {
+            PublishRef::Ssh(ref ssh_ref) => ssh_ref.url.as_str().to_owned(),
+            PublishRef::Https(ref https_ref) => https_ref.url.as_str().to_owned(),
+        }
+    }
+
+    /// return the a flakeref type for the wrapped refs
+    fn into_inner(self) -> FlakeRef {
+        match self {
+            PublishRef::Ssh(ssh_ref) => FlakeRef::GitSsh(ssh_ref),
+            PublishRef::Https(https_ref) => FlakeRef::GitHttps(https_ref),
+        }
+    }
+}
+
+impl TryFrom<FlakeRef> for PublishRef {
+    type Error = ConvertFlakeRefError;
+
+    fn try_from(value: FlakeRef) -> Result<Self, Self::Error> {
+        let publish_ref = match value {
+            FlakeRef::GitSsh(ssh_ref) => Self::Ssh(ssh_ref),
+            FlakeRef::GitHttps(https_ref) => Self::Https(https_ref),
+            // resolve upstream for local git repo
+            FlakeRef::GitPath(_) => todo!(),
+            // resolve indirect ref to direct ref (recursively)
+            FlakeRef::Indirect(_) => todo!(),
+            FlakeRef::Github(_) => todo!(),
+            FlakeRef::Gitlab(_) => todo!(),
+            _ => Err(ConvertFlakeRefError::UnsupportedTarget(value))?,
+        };
+        Ok(publish_ref)
+    }
+}
+
+/// Errors arising from convert
+#[derive(Error, Debug)]
+pub enum ConvertFlakeRefError {
+    #[error("Unsupported flakeref for publish: {0}")]
+    UnsupportedTarget(FlakeRef),
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use crate::flox::tests::flox_instance;
+    use crate::prelude::Channel;
+
+    #[cfg(feature = "impure-unit-tests")] // disabled for offline builds, TODO fix tests to work with local repos
+    #[tokio::test]
+    async fn creates_catalog_entry() {
```

Comment:
This isn't actually asserting anything, is it?

Reply:
No in this form not really, except that "no error occurred" which admittedly doesnt say that much 

1. it uses `git+ssh://git@github.com/flox/flox` and `github:flox/nixpkgs/stable` _unlocked_ flakes so the expected result will keep changing.
2. I wanted to at least assert the format is correct using flox types put that's blocked on them needing to integrate runix' FlakeRefs
3. this should be made runnable in offline modes (i.e. during nix build) to be meaningful, which means we have to pin nixpkgs and a target repo.. (which is where that `PublishRef::File` type from above would become relevant.